File: backend/app/schemas/assessment.py
import uuid
from datetime import datetime
from decimal import Decimal
from pydantic import BaseModel


class QuestionOptionCreate(BaseModel):
    text: str
    is_correct: bool = False


class QuestionOptionRead(BaseModel):
    id: uuid.UUID
    text: str
    is_correct: bool | None = None  # hidden from students

    model_config = {"from_attributes": True}


class QuestionCreate(BaseModel):
    text: str
    question_type: str  # mcq | true_false | short_answer | essay
    points: Decimal = Decimal("1.0")
    position: int = 0
    explanation: str | None = None
    options: list[QuestionOptionCreate] = []


class QuestionRead(BaseModel):
    id: uuid.UUID
    text: str
    question_type: str
    points: Decimal
    position: int
    explanation: str | None
    options: list[QuestionOptionRead]

    model_config = {"from_attributes": True}


class QuizCreate(BaseModel):
    title: str
    instructions: str | None = None
    time_limit_min: int | None = None
    max_attempts: int = 1
    shuffle_questions: bool = False
    due_at: datetime | None = None
    lesson_id: uuid.UUID | None = None


class QuizUpdate(BaseModel):
    title: str | None = None
    instructions: str | None = None
    time_limit_min: int | None = None
    max_attempts: int | None = None
    shuffle_questions: bool | None = None
    due_at: datetime | None = None
    is_published: bool | None = None


class QuizRead(BaseModel):
    id: uuid.UUID
    course_id: uuid.UUID
    title: str
    instructions: str | None
    time_limit_min: int | None
    max_attempts: int
    due_at: datetime | None
    is_published: bool
    question_count: int = 0
    total_points: Decimal = Decimal("0")
    created_at: datetime

    model_config = {"from_attributes": True}


class QuizDetailRead(QuizRead):
    questions: list[QuestionRead]


class AnswerSubmit(BaseModel):
    question_id: uuid.UUID
    selected_option_id: uuid.UUID | None = None
    text_response: str | None = None


class SubmissionCreate(BaseModel):
    answers: list[AnswerSubmit]


class AnswerRead(BaseModel):
    id: uuid.UUID
    question_id: uuid.UUID
    selected_option_id: uuid.UUID | None
    text_response: str | None
    is_correct: bool | None
    points_earned: Decimal | None
    feedback: str | None

    model_config = {"from_attributes": True}


class SubmissionRead(BaseModel):
    id: uuid.UUID
    quiz_id: uuid.UUID
    student_id: uuid.UUID
    attempt_num: int
    started_at: datetime
    submitted_at: datetime | None
    score: Decimal | None
    status: str
    answers: list[AnswerRead] = []

    model_config = {"from_attributes": True}


class ManualGradeRequest(BaseModel):
    answer_id: uuid.UUID
    points_earned: Decimal
    feedback: str | None = None

File: frontend/src/types/assessment.ts
export interface QuestionOption {
  id: string
  text: string
  is_correct?: boolean | null
}

export interface Question {
  id: string
  text: string
  question_type: 'mcq' | 'true_false' | 'short_answer' | 'essay'
  points: number
  position: number
  explanation: string | null
  options: QuestionOption[]
}

export interface Quiz {
  id: string
  course_id: string
  title: string
  instructions: string | null
  time_limit_min: number | null
  max_attempts: number
  due_at: string | null
  is_published: boolean
  question_count: number
  total_points: number
  created_at: string
  questions?: Question[]
  attempts_used?: number | null
}

export interface Answer {
  id: string
  question_id: string
  selected_option_id: string | null
  text_response: string | null
  is_correct: boolean | null
  points_earned: number | null
  feedback: string | null
}

export interface Submission {
  id: string
  quiz_id: string
  student_id: string
  attempt_num: number
  started_at: string
  submitted_at: string | null
  score: number | null
  status: 'in_progress' | 'submitted' | 'graded'
  answers: Answer[]
}
